br/pkg/restore/snap_client/pitr_collector.go

```diff
@@ -0,0 +1,472 @@
+		zap.Stringer("uuid", c.restoreUUID), zap.String("path", c.metaPath()), zap.Stringer("takes", time.Since(begin)))
+	return nil
+}
+
+func (c *pitrCollector) persistExtraBackupMeta(ctx context.Context) (err error) {
+	return c.writerRoutine.write(ctx)
+}
+
+// Commit commits the collected SSTs to a migration.
+func (c *pitrCollector) prepareMig(ctx context.Context) error {
+	if !c.enabled {
+		return nil
+	}
+
+	est := stream.MigrationExtension(c.taskStorage)
+
+	m := stream.NewMigration()
+	m.IngestedSstPaths = append(m.IngestedSstPaths, c.metaPath())
+
+	_, err := est.AppendMigration(ctx, m)
+	if err != nil {
+		return errors.Annotatef(err, "failed to add the extra backup at path %s", c.metaPath())
+	}
+
+	c.doWithMetaLock(func() {
+		c.resetCommitting()
+	})
+	// Persist the metadata in case of SSTs were uploaded but the meta wasn't,
+	// which leads to a leakage.
+	return c.persistExtraBackupMeta(ctx)
+}
+
+func (c *pitrCollector) prepareMigIfNeeded(ctx context.Context) (err error) {
+	c.putMigOnce.Do(func() {
+		err = c.prepareMig(ctx)
+	})
+	return
+}
+
+func (c *pitrCollector) commit(ctx context.Context) (uint64, error) {
+	c.extraBackupMeta.msg.Finished = true
+	ts, err := c.tso(ctx)
+	if err != nil {
+		return 0, err
+	}
+	c.extraBackupMeta.msg.AsIfTs = ts
+	return ts, c.persistExtraBackupMeta(ctx)
+}
+
+func (c *pitrCollector) resetCommitting() {
+	c.extraBackupMeta = ingestedSSTsMeta{
+		rewrites: map[int64]int64{},
+	}
+	c.extraBackupMeta.msg.FilesPrefixHint = c.sstPath("")
+	c.extraBackupMeta.msg.Finished = false
+	c.extraBackupMeta.msg.BackupUuid = c.restoreUUID[:]
+}
+
+// PiTRCollDep is the dependencies of a PiTR collector.
+type PiTRCollDep struct {
+	PDCli   pd.Client
+	EtcdCli *clientv3.Client
+	Storage *pb.StorageBackend
+}
+
+// newPiTRColl creates a new PiTR collector.
+func newPiTRColl(ctx context.Context, deps PiTRCollDep) (*pitrCollector, error) {
+	mcli := streamhelper.NewMetaDataClient(deps.EtcdCli)
+	ts, err := mcli.GetAllTasks(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(ts) > 1 {
+		return nil, errors.Annotatef(berrors.ErrInvalidArgument, "more than one task found, pitr collector doesn't support that")
+	}
+	if len(ts) == 0 {
+		return &pitrCollector{}, nil
+	}
+
+	coll := &pitrCollector{
+		enabled: true,
+	}
+
+	strg, err := storage.Create(ctx, ts[0].Info.Storage, false)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	coll.taskStorage = strg
+
+	tso := func(ctx context.Context) (uint64, error) {
+		l, o, err := deps.PDCli.GetTS(ctx)
+		return oracle.ComposeTS(l, o), err
+	}
+	coll.tso = tso
+
+	t, err := tso(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	coll.name = fmt.Sprintf("backup-%016X", t)
+
+	restoreStrg, err := storage.Create(ctx, deps.Storage, false)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	coll.restoreStorage = restoreStrg
+	coll.restoreSuccess = summary.Succeed
+	coll.goPersister()
+	coll.resetCommitting()
```

Comment:
not big issue, but if enable==false, this goPersister will leak.
and resetCommitting seems duplicated with PerpareMig

Reply:
If there is no task, we will return eariler:

<img width="734" alt="image" src="https://github.com/user-attachments/assets/08aa2f36-96a8-485e-9ffa-ea76d5071920" />

Reply:
There is a task, but enable==false

Reply:
if enable==false, it won't return eariler.

Reply:
It is an internal function, I think we can control don't call it when `enabled` is false...

Reply:
Anyway we can modify `close` to make it always close the background goroutine to be defensive.